Change specifications according to consensus-application separation

janhack opened the pull request:
The BFT module is not implemented as a module any more, but instead moved to the engine. We therefore update the application layer in order to remove any information flow from engine to state machine. This implies the following changes:

## Update of LIP 0044

- Removed the rationale section about block slots and moved it to LIP 0058 instead.
- The Validators module now stores the certificate threshold, precommit threshold and for every validator in a round the address, BFT weight, generator key and BLS key. This way the Validators module can forward this information to the engine and other modules can request this information for the current round if needed.
- The Generator List Substore is renamed to the Validators Params Substore which stores the information mentioned above.
- The function to set and get the generator list is renamed to `setValidatorParams` and `getValidatorParams` and generalized.
- The function `getSlotTime` is removed as it is not used any more.
- The functions `getGeneratorAtTimestamp` and `getSlotNumber` are moved to LIP 0058 as they are used in the engine during the block processing and not used in the state machine any more.
- Remove endpoint `getValidatorParams` as this information will be provided by the engine.

## Update of LIP 0050

- Change Validator method call from `getValidatorAccount` to `getValidatorKeys`. 

## Update of LIP 0055

- We add the property `impliesMaxPrevotes` to the block header. This property will be validated by the engine and can be used by the Rewards module to determine whether the block reward should be reduced. Therefore the call to the method `impliesMaximalPrevotes` is not needed any more.
- Update references to the functions `getGeneratorAtTimestamp` and `getSlotNumber`, which are now defined in LIP 0058.

## Update of LIP 0056

- Update links.
- Remove any reference to "BFT module".

## Update of LIP 0058

- Added the section about block slots that was previously in LIP 0044.
- Removed any reference that the BFT logic is implemented as a module.
- Use terminology "consensus domain" and "application domain" instead of replication layer or state machine.
- Add Genesis Data Substore for storing the genesis block timestamp.
- The function `setBFTParameters` is changed as follows:
    - The array `validators` given as input is not sorted and stored in the BFT Parameters store in unchanged order. This array determines the block generation list.
    - The BFT weight of the validators is allowed to be 0.
    - Any calls to the Validators module can be removed as the `validators` array already contains the generator and BLS keys. 
- Fix function `impliesMaximalPrevotes` to work for the first block after the genesis block.
- Remove function `getCurrentValidators` as it is not needed any more.
- Add function `getSlotNumber` (moved from LIP 0044)
- Add function `getGeneratorAtTimestamp` (moved from LIP 0044)

## Update of LIP 0059

- Use the property `b.header.aggregateCommit.height` instead of requesting the current value of `maxHeightCertified` from the BFT store.

## Update of LIP 0060

- Add `b.header.impliesMaxPrevotes == True` to static validation.
- Add `eventRoot` property to result verification stage.
- Change `assetsRoot` to `assetRoot`.

ricott1 commented on proposals/lip-0044.md:
```diff
@@ -127,29 +130,76 @@ The validator account holds the generator and BLS keys of a registered validator
-* The store value is set to the _generator list_ data structure, serialized using the JSON schema `generatorListSchema`, presented [below](#json-schema-1).
-* Notation: For the rest of this proposal let `generatorList` be the entry in the generator list substore (for notational convenience, we refer directly to the array of addresses, without explicitly writing `generatorList.addresses`).
+* The store value is the serialization of an object following the JSON schema `validatorParamsSchema` defined below.
+* Notation: For the rest of this proposal let `validatorsParams` be the entry in the generator list substore.
```
"... in the validator params substore."

janhack replied:
Fixed.

janhack commented on proposals/lip-0044.md:
```diff
@@ -596,29 +646,71 @@ def getGeneratorsBetweenTimestamps(startTimestamp: uint32, endTimestamp: uint32)
-        if no entry in the validators data substore exist with storeKey == address:
-            raise Exception(f'Invalid address in generator list. The address {address.hex()} is not registered as validator.')
-    generatorList = generatorAddresses
+setValidatorParams(precommitThreshold: uint64, certificateThreshold: uint64, validatorList: list[object]):
```
I tried to put the function from the BFT module here 1-1 and also store exactly the same information as in the BFT Params store, but of course we could change it. I was thinking it would be nice to do the validation here so if in the application layer this method is called with invalid parameters, an exception is raised directly. Additionally, I though it is nice to have all consensus-related properties here in one place as they can then be also requested via an endpoint. However, in terms of protocol it would not be necessary to store them. I am open to suggestions here.

shuse2 replied:
I think it would be nicer if we can avoid validators module to know too much about BFT protocol, so I would prefer to keep it as proxy to reply to the engine which stores the value as it is.

Validation wise it will be the same in most cases. Even if we don't throw in the application, it will be thrown in the engine and the block will be invalid.
Maybe only case where it could be different is if there is integration test without engine, developer will not notice, but I think that's probably true for other replies for engine (like invalid event schema etc), also this can be resolved by mock ABI implementation.

Currently, there is `consensus` endpoint exposed from the engine, so BFT related properties can be fetched from the endpoint (for certificate etc)

janhack replied:
Okay, then in terms of specifications, the function `setValidatorParams` is a simple getter without validation, but in the BFT LIP I keep the full function with full validation. I will also remove the `prevoteThreshold` and `validatorsHash` property from the store.

janhack replied:
This comment can be resolved from my side.

janhack commented on proposals/lip-0055.md:
```diff
@@ -599,16 +609,22 @@ def verifyEventRoot():
+  if bft.isHeaderContradictingChain(block.header):
+    return False
+
+  if block.header.impliesMaxPrevotes != bft.impliesMaximalPrevotes(block.header)
```
Isn't the whole function `verifyBFTProperties()` part of the engine now so all information would be available?

shuse2 replied:
yes, all the information is available. I guess I just need to slightly modify the `impliesMaximalPrevotes` function to handle genesis block case, but it can be handled.

ricott1 commented on proposals/lip-0044.md:
```diff
@@ -127,32 +124,69 @@ The validator account holds the generator and BLS keys of a registered validator
+            "fieldNumber": 2
+        },
+        "generatorKey": {
+            "dataType": "uint64",
```
```suggestion
            "dataType": "bytes",
```

janhack replied:
Fixed.

ricott1 commented on proposals/lip-0059.md:
```diff
@@ -41,16 +41,16 @@ lastHeightOfRound(h + 2 * ROUND_LENGTH) <= maxHeightCertified.
-- We use the constant `ROUND_LENGTH` defined in the [LIP "Define state and state transitions of DPoS module"][research:dpos-module] to denote the round length in the DPoS chain, i.e., the number of active plus randomly selected standby delegates.
-- We write `bftModule.fct` for a call to the function `fct` defined in the [BFT module][research:bft-module].
+- We use the constant `ROUND_LENGTH` defined in [LIP 0057][lip-0057] to denote the round length in the DPoS chain, i.e., the number of active plus randomly selected standby delegates.
+- We write `bftModule.fct` for a call to the function `fct` defined in the [BFT module][lip-0058].
```
We should not refer to it as a module, right?

janhack replied:
Fixed, I removed this bullet point.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -626,26 +674,50 @@ The function allows to obtain the next larger key in the BFT Parameters substore
 #### setBFTParameters
 
-This function allows to set the thresholds, validators and associated BFT weights to be used from the next height onward and updates the BFT module store accordingly.
+This function allows to set the thresholds, validators and associated BFT weights to be used from the next height onward and updates the BFT module store accordingly. The function should be called with the parameters forwarded to the application layer from the state machine every time the function [setValidatorParams][lip-0044#setValidatorParams] of the Validators module is called.
```
```suggestion
This function allows to set the thresholds, validators and associated BFT weights to be used from the next height onward and updates the BFT store accordingly. The function should be called with the parameters forwarded to the application layer from the state machine every time the function [setValidatorParams][lip-0044#setValidatorParams] of the Validators module is called.
```

janhack replied:
Fixed.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
 ## Motivation
 
-The Lisk-BFT consensus protocol introduced in [LIP 0014](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0014.md) specifies two new block properties, `maxHeightGenerated` (renamed from `maxHeightPreviouslyForged`) and `maxHeightPrevoted`, which are validated as part of the block processing. As part of the new state architecture and stricter separation between state machine layer and replic ...
+The Lisk-BFT consensus protocol introduced in [LIP 0014](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0014.md) specifies two new block properties, `maxHeightGenerated` (renamed from `maxHeightPreviouslyForged`) and `maxHeightPrevoted`, which are validated as part of the block processing. As part of the [new state architecture][lip-0069] and stricter separation between state machine laye ...
```
I would also mention the property `impliesMaximalPrevotes` (although it is not introduced in LIP14), since it is handled by the BFT store

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
 * **Block headers information**: A subset of the properties of the recent block headers in the current chain is required to update the prevote and precommits weights. The reason is that validators cast prevotes and precommits as part of their block header, see [LIP 0056][lip-0056] for details.
 
-We call the first four inputs above the **BFT parameters**. For the BFT module, we assume that these BFT parameters are provided from another module via the `setBFTParameters` function defined below. For a DPoS chain, the DPoS module would need to call the BFT module with the correct parameters, in particular, notifying the BFT module about any changes of validators. Similarly, for a PoA chain, i ...
+We call the first four inputs above the **BFT parameters**. We assume that these BFT parameters are forwarded from the state machine layer to the replication layer whenever the [setValidatorParams][lip-0044#setValidatorParams] function in the Validators module of the state machine layer is called. For a DPoS chain, the DPoS module would need to call this function with the correct parameters, in p ...
```
I have some problems parsing the last sentence. I would find it clearer to just say smtg like: "The block header information is always available to the replication layer." or even be more specific about what is considered the 'engine-state'.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -268,9 +292,29 @@ bftVotesSchema = {
+* The substore prefix is set to `SUBSTORE_PREFIX_GENESIS_DATA`.
+* The store key is set to empty bytes.
+* The store value is set to a data structure holding the timestamp of the genesis block of the chain, serialized using the `genesisDataSchema` schema presented below.
+* Notation: For the rest of this proposal let `genesisData` be the entry in the genesis data substore.
```
you are a bit inconsistent with the capitalization of substore names in the LIP

janhack replied:
Changed to "Genesis Data substore" here and in other places in LIP 58 for consistency.

ricott1 replied:
I think you still have inconsistencies in the LIP

ricott1 replied:
I was looking at the outdated version, resolving.

ricott1 commented on proposals/lip-0058.md:
````diff
@@ -401,7 +445,7 @@ The BFT parameters valid at the height `h`.
 
 ```python
-getBFTParametersInternal(h):
+def getBFTParametersInternal(h):
````
Here and later, I would still add the typing

janhack replied:
I would prefer not to do it as I wanted to restrict the PR to necessary specification changes, we have all the "Parameters" section which already describe the input parameters, it would not help to save additional validations and there are also a lot of complex objects as input parameters. I think we also said that introducing type hinting for published LIPs is optional.

ricott1 replied:
sounds good.

ricott1 commented on proposals/lip-0058.md:
````diff
@@ -487,7 +531,7 @@ The function returns `True` if and only if there is an entry in the BFT Paramete
 
 ```python
-existBFTParameters(h):
+def existBFTParameters(h):
````
Does `h` represent an height here? what type is it? if it is uint we still need to convert it internally

janhack replied:
The input parameters is described in the section Parameters above. Changed the line below to
```
if exists entry in BFT Parameters store with key uint32be(h):
```

ricott1 commented on proposals/lip-0058.md:
````diff
@@ -487,7 +531,7 @@ The function returns `True` if and only if there is an entry in the BFT Paramete
 ```python
-existBFTParameters(h):
+def existBFTParameters(h):
     if exists entry in BFT Parameters store with key h:
````
```suggestion
    if exists entry in BFT Parameters store with store key h:
```

janhack replied:
Fixed.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -694,29 +765,28 @@ setBFTParameters(precommitThreshold, certificateThreshold, validators):
+In this step, the function [setValidatorParams][lip-0044#setValidatorParams] of the Validators module should be called by the PoA module or DPoS module.
+These parameters are forwarded to the replication layer and the BFT parameters store is updated accordingly using the function `setBFTParameters` defined above.
 
 ### Block Processing
```
I think this section should be removed. As this is not a module anymore, the state machine hooks would not be called. If the logic is still processed, we should clarify when that happens (probably in the replication layer).

janhack replied:
As concluded in our discussions, I now have a section "Before Application Processing" and "After Application Processing" for the genesis block processing.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -736,21 +806,23 @@ Before the transactions of a block `b` are executed, the following logic is exec
 
 ### Block Creation
```
Also here, it is not clear to me when would this logic be called.

janhack replied:
I improved this section and reference the concrete block generation stage defined in LIP 55. Note that as far as I understand both the header initialization and header finalization stage are in the engine domain.

ricott1 replied:
I think we can still improve by referencing exactly in which stage this happens.

ricott1 replied:
same problem, now it's fine, resolving.

ricott1 reviewed (changes requested):
In general, it looks good. I still have some comments, in particular related to the (ex)BFT module hooks.

AndreasKendziorra commented on proposals/lip-0044.md:
```diff
@@ -127,32 +124,69 @@ The validator account holds the generator and BLS keys of a registered validator
-* The store value is set to the _generator list_ data structure, serialized using the JSON schema `generatorListSchema`, presented [below](#json-schema-1).
-* Notation: For the rest of this proposal let `generatorList` be the entry in the generator list substore (for notational convenience, we refer directly to the array of addresses, without explicitly writing `generatorList.addresses`).
+* The store value is the serialization of an object following the JSON schema `validatorParamsSchema` defined below.
+* Notation: For the rest of this proposal let `validatorsParams` be the entry in the validators params substore.
```
sometimes it's "validatorsParams" and sometimes "validatorParams"

janhack replied:
It should now be "validatorParams" everywhere.

AndreasKendziorra commented on proposals/lip-0059.md:
````diff
@@ -61,33 +61,33 @@ Here
 ```
 
-Here `roundNumber` is a function defined in the [LIP "Define state and state transitions of DPoS module"][research:dpos-module#roundnumber] and `genesisHeight` denotes the height of the genesis block as stored in the [DPoS module][research:dpos-module].
+Here `roundNumber` is a function defined in [LIP 0057][lip-0057#roundnumber] and `genesisDataStore.height` denotes the height of the genesis block as stored in the [DPoS module][lip-0057].
````
> `roundNumber` is a function defined in [LIP 0057]

this is maybe a bit misleading because this function is not public. So maybe we should define this internal function here as well or just replace `roundNumber(h)` by the formula.

janhack replied:
This LIP defines some additional functionality of the DPoS module and would therefore be implemented as part of the DPoS module. Therefore, I assume here that I can use internal functions of the DPoS module.

AndreasKendziorra reviewed (changes requested):
The term "BFT module" is still used in the LIPs 0042, 0047, 0056, 0057, 0061

janhack commented on proposals/lip-0058.md:
```diff
@@ -626,26 +674,50 @@ The function allows to obtain the next larger key in the BFT Parameters substore
 #### setBFTParameters
 
-This function allows to set the thresholds, validators and associated BFT weights to be used from the next height onward and updates the BFT module store accordingly.
+This function allows to set the thresholds, validators and associated BFT weights to be used from the next height onward and updates the BFT module store accordingly. The function should be called with the parameters forwarded to the application layer from the state machine every time the function [setValidatorParams][lip-0044#setValidatorParams] of the Validators module is called.
```
Updated to

> The function is called in the after application processing stage of the genesis block processing and general block processing as defined below.

In this LIP I use the more general academic terminology (state machine, replication layer) instead of the concrete Lisk SDK architecture terminology (application domain, engine domain). Not sure if this is the best approach, I will raise this point on Slack.

janhack replied:
As concluded on Slack, I now only use the terms application domain and consensus domain as introduced in LIP 69.

janhack commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
 ## Motivation
 
-The Lisk-BFT consensus protocol introduced in [LIP 0014](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0014.md) specifies two new block properties, `maxHeightGenerated` (renamed from `maxHeightPreviouslyForged`) and `maxHeightPrevoted`, which are validated as part of the block processing. As part of the new state architecture and stricter separation between state machine layer and replic ...
+The Lisk-BFT consensus protocol introduced in [LIP 0014](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0014.md) specifies two new block properties, `maxHeightGenerated` (renamed from `maxHeightPreviouslyForged`) and `maxHeightPrevoted`, which are validated as part of the block processing. As part of the [new state architecture][lip-0069] and stricter separation between state machine laye ...
```
Fixed both points. I would keep "logic" as not only verification is performed in the replication layer.

janhack commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
 * **Block headers information**: A subset of the properties of the recent block headers in the current chain is required to update the prevote and precommits weights. The reason is that validators cast prevotes and precommits as part of their block header, see [LIP 0056][lip-0056] for details.
 
-We call the first four inputs above the **BFT parameters**. For the BFT module, we assume that these BFT parameters are provided from another module via the `setBFTParameters` function defined below. For a DPoS chain, the DPoS module would need to call the BFT module with the correct parameters, in particular, notifying the BFT module about any changes of validators. Similarly, for a PoA chain, i ...
+We call the first four inputs above the **BFT parameters**. We assume that these BFT parameters are forwarded from the state machine layer to the replication layer whenever the [setValidatorParams][lip-0044#setValidatorParams] function in the Validators module of the state machine layer is called. For a DPoS chain, the DPoS module would need to call this function with the correct parameters, in p ...
```
- I removed the part "notifying the replication layer about any changes of validators".
- Changed to "input to the setValidatorParams function" as suggested.
- As concluded on Slack, I now only use the terms application domain and consensus domain as introduced in LIP 69.
- The DPoS module should already be updated on LIPs staging. For the PoA module, I would suggest that Sergey includes this change as he is already a PR for updating the module. I will send a message to Sergey about this point.

janhack commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
+
+### Block Slots
+
+In the Lisk protocol, time is divided into intervals of fixed length. These intervals are called _block slots_ and at most one block can be added to the blockchain during each slot. The length of a block slot is a constant called _block time_. In this LIP we define how to determine whether which generator is eligible to generate a block in a certain block slot.
```
Fixed.

janhack commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
 | `MAX_LENGTH_BLOCK_BFT_INFOS`  |`uint32`  | `3*LSK_BFT_BATCH_SIZE` | The maximum length of the array `blockBFTInfos` in the BFT Votes substore. |
+| `BLOCK_TIME`                  |`uint32`  | 10 (value for the Lisk mainchain) | Block time (in seconds) set in the chain configuration.         |
+| `ADDRESS_LENGTH`              |`uint32`  | 20                     | Length in bytes of an address.                                             |
+| `ED25519_PUBLIC_KEY_LENGTH`   |`uint32`  | 32                     | Length in bytes of a Ed25519 public keys.                                  |
```
Fixed.

janhack commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
+| `BLOCK_TIME`                  |`uint32`  | 10 (value for the Lisk mainchain) | Block time (in seconds) set in the chain configuration.         |
+| `ADDRESS_LENGTH`              |`uint32`  | 20                     | Length in bytes of an address.                                             |
+| `ED25519_PUBLIC_KEY_LENGTH`   |`uint32`  | 32                     | Length in bytes of a Ed25519 public keys.                                  |
+| `BLS_PUBLIC_KEY_LENGTH`       |`uint32`  | 48                     | Length in bytes of BLS public key.                                         |
```
Fixed.

janhack commented on proposals/lip-0044.md:
```diff
@@ -594,32 +595,45 @@ def getGeneratorsBetweenTimestamps(startTimestamp: uint32, endTimestamp: uint32)
+    params.validators = newValidators
+
+    # write new validator params to state store
+    validatorParams = params
```
My suggestion would then be to use "validatorsParamsStore" for the state store notation, i.e., always have "store" as part of the name.

janhack commented on proposals/lip-0058.md:
```diff
@@ -74,18 +82,18 @@ As explained in [Section "Comparison to the Lisk-BFT paper" in LIP 0056][lip-005
 
 #### BFT Parameters
 
 ##### Store Prefix, Store Key, and Store Value
```
- I added the sentence "Each substore of the BFT store is a simple key-value store similar to the substores in the application domain."
- I agree that we don't necessarily need these exact substore prefixes, but we would still need different prefixes for different substores. I think it is also fine to define them in the LIP although the protocol would be the same. @shuse2 Can you give more details on how the substore prefixes are used?

AndreasKendziorra replied:
Fine for me. I would leave it to @shuse2 if more changes are needed here.

shuse2 replied:
prefix are used to split the data in the database, so i think it's good to keep it there.
how it's currently structured is

`[STATE_PREFIX, SUB_STORE_PREFIX, KEY]` similar to state db in the application

janhack replied:
Thanks for the clarification. I think this comment can be resolved then.

janhack commented on proposals/lip-0058.md:
```diff
@@ -136,6 +133,33 @@ bftParametersSchema = {
+            "fieldNumber": 2
+        },
+        "generatorKey": {
+            "dataType": "uint64",
```
Fixed.

janhack commented on proposals/lip-0058.md:
```diff
@@ -136,6 +133,33 @@ bftParametersSchema = {
+            "fieldNumber": 3
+        },
+        "blsKey": {
+            "dataType": "uint64",
```
Fixed.

janhack commented on proposals/lip-0058.md:
```diff
@@ -147,17 +171,17 @@ All properties below are valid starting from the height given as key:
 
-* The store prefix is `STORE_PREFIX_BFT_VOTES`.
+* The substore prefix is `SUBSTORE_PREFIX_BFT_VOTES`.
 * The store key is empty bytes.
```
As far as I know we only have one persistent storage that we use in the Lisk SDK, the RocksDB database which is basically one giant key-value store. I think for consistency it makes sense to also store this information there and to distinguish it from other substores, it makes sense to have a certain prefix.

janhack commented on proposals/lip-0058.md:
```diff
@@ -546,7 +596,7 @@ The function returns `True` if the block header `blockHeader` provided as input
 ##### Execution
```
Changed to "Update ... to the new _store_ structure".

janhack commented on proposals/lip-0044.md:
```diff
@@ -594,32 +595,45 @@ def getGeneratorsBetweenTimestamps(startTimestamp: uint32, endTimestamp: uint32)
+#### setValidatorParams
 
-This function updates the value of the generator list to the input `generatorAddresses`. It checks if all elements of `generatorAddresses` have a registered validator account in the validators data substore. If this is not the case, it throws an error. Otherwise, it updates the value of `generatorList`.
+This function allows to set the precommit threshold, certificate threshold, validators and associated BFT weights to be used from the next height onward. The information is not validated and simply stored in the Validators module. The information is further forwarded to the engine where all validations are performed.
```
I added the following line to the function:
```
forward params to consensus domain
```

What happens in the engine if the validation fails, is now defined in the section "After Application Processing" in LIP 58.

AndreasKendziorra replied:
ok. It would be good to reference the sections from LIP 58 as it's very hard for a reader of this section to see where in LIP 58 this data is processed /validated.
Maybe it's also not very obvious that the consensus domain has to keep the forwarded data in memory until it a certain stage (after application processing) is reached and will not process/validate it directly.

One idea: LIP 0058 introduces an auxiliary data structure `aux_BFT_params` that is kept in memory and that is filled by a function `set_aux_BFT_params` which is called by `setValidatorParams`. Then in  After Application Processing `setBFTParameters`is called with `aux_BFT_params` if `aux_BFT_params` not empty. At the end of After Application Processing is cleared such that it is empty once the next block is processed.

janhack replied:
I added some links to the function description to make it more clear what happens.

I would not add any auxiliary data structure for temporary storage or get very detailed regarding the implementation. @shuse2  was explaining that the transfer of information basically works as follows:
```
const result = abi.afterTransactionsExecute();
```
If I understand correctly here the application is called from the consensus domain via the application-blockchain-interface. What "forward params to consensus domain" is to add the BFT parameters to the return value of the function `afterTransactionsExecute`. The `result` value is then used in the after application processing stage within the consensus domain.

janhack commented on proposals/lip-0058.md:
````diff
@@ -626,26 +674,50 @@ The function allows to obtain the next larger key in the BFT Parameters substore
+    return floor(elapsedTime / BLOCK_TIME)
+```
+
+#### getValidator
````
I removed the function.

janhack commented on proposals/lip-0058.md:
```diff
@@ -694,29 +765,28 @@ setBFTParameters(precommitThreshold, certificateThreshold, validators):
+  * `bftVotes.maxHeightCertified = b.header.height`,
   * `bftVotes.blockBFTInfos` is an empty array,
   * `bftVotes.activeValidatorsVoteInfo` is an empty array.
+* The Genesis Data substore is initialized by setting `genesisData.timestamp = h.header.timestamp`.
```
Fixed.

janhack commented on proposals/lip-0044.md:
````diff
@@ -511,46 +545,13 @@ def getAddressFromBLSKey(blsKey: PublicKeyBLS) -> Address:
-```
-
-#### getSlotNumber
+#### getValidatorParams
````
This is introduced so that other modules have the flexibility to obtain this information from the Validators module. However, for our modules we don't need it so far.

janhack commented on proposals/lip-0044.md:
````diff
@@ -395,15 +429,15 @@ def registerValidatorWithoutBLSKey(validatorAddress: Address, generatorKey: Publ
 ```
 
-#### getValidatorAccount
+#### getValidatorKeys
````
I fixed the 1 occurrence in LIP 55 and the 1 occurrence in LIP 50. For all other LIPs there is an in-progress LIP update and I added a point on the Trello card "Follow up tasks after engine-application separation".

janhack commented on proposals/lip-0058.md:
````diff
@@ -591,27 +641,25 @@ The BFT parameters valid at the height `h`.
 
 ```python
-getCurrentValidators()
+def getGeneratorAtTimestamp(timestamp):
````
I added an additional input parameter "currentHeight" and updated the function description.

janhack commented on proposals/lip-0058.md:
```diff
@@ -655,15 +727,14 @@ setBFTParameters(precommitThreshold, certificateThreshold, validators):
-        activeValidator.blsKey =  validatorModule.getValidatorAccount(validator.address).blsKey
+        activeValidator.blsKey =  validator.blsKey
         add activeValidator to activeValidators
     sort elements in activeValidators lexicographically by blsKey property
```
Very good point. I added the check.

janhack commented:
> Should we also update LIP0042 (reward) to use the block heade impliesMaximalPrevotes property in https://github.com/LiskHQ/lips/blob/main/proposals/lip-0042.md#execution ?

@ricott1 I also fixed this point.

> The term "BFT module" is still used in the LIPs 0042, 0047, 0056, 0057, 0061

@AndreasKendziorra I removed the references in LIP 42 and 56. For the LIPs 47, 57 and 61 there is a pending update so I did not change it here, but instead added these points to a Trello card.

janhack commented:
I addressed all review comments and re-requested your reviews.

AndreasKendziorra commented on proposals/lip-0044.md:
````diff
@@ -127,32 +124,69 @@ The validator account holds the generator and BLS keys of a registered validator
+}
 ```
 
 ##### Properties and Default values
````
this was overlooked?

janhack replied:
Sorry, missed it. I fixed this section and moved the reference to `getGeneratorAtTimestamp` to the parameter description of the function `setValidatorParams`.

ricott1 commented on proposals/lip-0042.md:
```diff
@@ -206,6 +206,5 @@ TBA
 [creative]: https://creativecommons.org/publicdomain/zero/1.0/
 [block-schema-lip]: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md
 [bft-module]: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0058.md
```
I think this link can be removed

janhack replied:
Done.

ricott1 reviewed (changes requested):
Iresolved most of my comments, an extra one:

- We still need to add `impliesMaxPrevotes` to the section "New Block Header Properties" of LIP0055.

ricott1 commented on proposals/lip-0060.md:
```diff
@@ -41,17 +41,21 @@ Another distinctive specification of the genesis block is its processing. The bl
 
```
For consistency we could also just name this section "Static validation"

ricott1 replied:
Also, in the previous paragraph we should amend "In particular the processing of the genesis block happens in **four** steps described in the sections below." --> "In particular the processing of the genesis block happens in **five** steps described in the sections below." (not sure if 'block storage should be included, if so, it should be 6)

janhack replied:
- I changed all occurrences of "Static validation of the genesis block" to "Static validation".
- Changed it from four to five steps.

ricott1 replied:
I would add the 'block storage' step to the description (you can copy/paste it from LIP 0055), since I put it in the diagram, and update from 'five' to 'six'.

janhack replied:
I added the block storage to the description in LIP 60.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
-| `MODULE_ID_BFT`               |`uint32`  | TBD                    | The module ID of the BFT module.                                           |
-| `STORE_PREFIX_BFT_PARAMETERS` |`bytes`   | `0x0000`               | The store prefix of the BFT Parameters substore.                           |
-| `STORE_PREFIX_BFT_VOTES`      |`bytes`   | `0x8000`               | The store prefix of the BFT Votes substore.                                |
+| `SUBSTORE_PREFIX_BFT_PARAMETERS` |`bytes`   | `0x0000`               | The store prefix of the BFT Parameters substore.  |
```
It should be 'substore prefix'

janhack replied:
Fixed.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
-| `STORE_PREFIX_BFT_PARAMETERS` |`bytes`   | `0x0000`               | The store prefix of the BFT Parameters substore.                           |
-| `STORE_PREFIX_BFT_VOTES`      |`bytes`   | `0x8000`               | The store prefix of the BFT Votes substore.                                |
+| `SUBSTORE_PREFIX_BFT_PARAMETERS` |`bytes`   | `0x0000`               | The store prefix of the BFT Parameters substore.  |
+| `SUBSTORE_PREFIX_BFT_VOTES`   |`bytes`   | `0x4000`               | The store prefix of the BFT Votes substore.                                |
```
It should be 'substore prefix'

janhack replied:
Fixed.

ricott1 commented on proposals/lip-0058.md:
```diff
@@ -13,58 +13,66 @@ Requires: 0055, 0056, 0061
-| `STORE_PREFIX_BFT_VOTES`      |`bytes`   | `0x8000`               | The store prefix of the BFT Votes substore.                                |
+| `SUBSTORE_PREFIX_BFT_PARAMETERS` |`bytes`   | `0x0000`               | The store prefix of the BFT Parameters substore.  |
+| `SUBSTORE_PREFIX_BFT_VOTES`   |`bytes`   | `0x4000`               | The store prefix of the BFT Votes substore.                                |
+| `SUBSTORE_PREFIX_GENESIS_DATA`| bytes | `0x8000` | Substore prefix of the Genesis Data substore. |
```
```suggestion
| `SUBSTORE_PREFIX_GENESIS_DATA`| bytes | `0x8000` |  The substore prefix of the Genesis Data substore. |
```

janhack replied:
Changed as suggested.

janhack commented on proposals/lip-0044.md:
```diff
@@ -124,14 +124,14 @@ The validator account holds the generator and BLS keys of a registered validator
 
 ##### Substore prefix, Store Key, and Store Value
 
 * The substore prefix is set to `SUBSTORE_PREFIX_VALIDATORS_PARAMS`.
```
Good point, fixed.

janhack commented on proposals/lip-0058.md:
```diff
@@ -21,7 +21,7 @@ This LIP is licensed under the [Creative Commons Zero 1.0 Universal](https://cre
 ## Motivation
 
-The Lisk-BFT consensus protocol introduced in [LIP 0014](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0014.md) specifies two new block properties, `maxHeightGenerated` (renamed from `maxHeightPreviouslyForged`) and `maxHeightPrevoted`, which are validated as part of the block processing. As part of the [new state architecture][lip-0069] and stricter separation between state machine laye ...
+The Lisk-BFT consensus protocol introduced in [LIP 0014](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0014.md) specifies two block properties, `maxHeightGenerated` (renamed from `maxHeightPreviouslyForged`) and `maxHeightPrevoted`, which are validated as part of the block processing. As part of the [new state architecture][lip-0069] and stricter separation into network domain, consensus ...
```
Fixed, thanks.

janhack commented on proposals/lip-0058.md:
````diff
@@ -681,28 +694,16 @@ This function returns the slot number corresponding to the input timestamp.
-    raise Exception()
-```
-
 #### setBFTParameters
````
I would agree and moved the function to the section "Internal Functions". @shuse2 I hope this is correct.

AndreasKendziorra replied:
can be resolved from my side, but I leave it open for @shuse2 .

shuse2 replied:
it should be fine, I think it is used in consensus and generator but it's not open for any other modules

janhack commented on proposals/lip-0058.md:
```diff
@@ -561,7 +578,7 @@ def impliesMaximalPrevotes(blockHeader):
             return False
     heightCurrentTip = bftVotes.blockBFTInfos[0].height
```
I would not do more refactoring now to also keep this diff for this PR small and easier to implement for the developers.

janhack commented on proposals/lip-0060.md:
```diff
@@ -70,7 +74,7 @@ As introduced in [the Rationale section](#processing-of-the-genesis-block), the
 * _Static validation of the genesis block_: The stateless checks that ensure the structure of the genesis block are performed. Also, the properties in the block header that do not require access to the state store are checked. These checks are defined in the sections below.
 * _Genesis state initialization_: The genesis state initialization logic for the registered modules is executed.
```
Good catch, I added it.

shuse2 commented on proposals/lip-0058.md:
```diff
@@ -268,9 +292,29 @@ bftVotesSchema = {
   * `largestHeightPrecommit`: The largest height for which a block generated by the validator implied a precommit.
 
-### Commands
+#### Genesis Data Substore
```
I think it's fine to introduce it as notation and function.
Currently, genesis block is saved as any other blocks. At the beginning of starting the node, it checks if the genesis block exists and match with the genesis block given through the config.

there the genesis block header is passed to slot calculation object (Slot) and maintained in the consensus and used for syncing and forkchoice rule

janhack replied:
Okay. I removed the Genesis Data substore and added the two global variables as suggested above.

AndreasKendziorra commented on proposals/lip-0058.md:
```diff
@@ -664,7 +726,7 @@ def getBFTParameters(h):
 #### getGeneratorAtTimestamp
 
-This function returns the address of the generator active at the timestamp `timestamp` for the block at height `currentHeight` where `currentHeight` where `timestamp` and `currentHeight` are given as input. Notice that if the input timestamp corresponds to a time slot before or after the current round, this function may not return the correct generator address.
+This function returns the address of the generator active at the timestamp `timestamp` for the block at height `currentHeight + 1` where `currentHeight` where `timestamp` and `currentHeight` are given as input. Notice that if the input timestamp corresponds to a time slot before or after the current round, this function may not return the correct generator address.
```
```suggestion
This function returns the address of the generator active at the timestamp `timestamp` for the block at height `currentHeight + 1` where `timestamp` and `currentHeight` are given as input. Notice that if the input timestamp corresponds to a time slot before or after the current round, this function may not return the correct generator address.
```

janhack replied:
Fixed, thanks.

AndreasKendziorra commented on proposals/lip-0060.md:
```diff
@@ -31,13 +31,13 @@ The new genesis block format and processing is specified with the rationale of h
-Another distinctive specification of the genesis block is its processing. The blockchain starts with an empty key-value store and in particular, all module stores are initially empty. As part of the genesis block processing modules can add key-value entries to their module store to initialize their state. They should also check the consistency of their state independently and against information  ...
+Another distinctive specification of the genesis block is its processing. The blockchain starts with an empty key-value store and in particular, all module stores are initially empty. As part of the genesis block processing modules can add key-value entries to their module store to initialize their state. They should also check the consistency of their state independently and against information  ...
 
 <img src="lip-0060/stages.png" width="70%">
```
@ricott1 the legend is broken: no text next to the red square

janhack replied:
I removed concrete counts of the number of steps from this LIP.

janhack replied:
The diagram is also fixed.

AndreasKendziorra commented on proposals/lip-0060.md:
```diff
@@ -31,13 +31,13 @@ The new genesis block format and processing is specified with the rationale of h
 ### Processing of the genesis block
 
-Another distinctive specification of the genesis block is its processing. The blockchain starts with an empty key-value store and in particular, all module stores are initially empty. As part of the genesis block processing modules can add key-value entries to their module store to initialize their state. They should also check the consistency of their state independently and against information  ...
+Another distinctive specification of the genesis block is its processing. The blockchain starts with an empty key-value store and in particular, all module stores are initially empty. As part of the genesis block processing modules can add key-value entries to their module store to initialize their state. They should also check the consistency of their state independently and against information  ...
```
> in five steps described in the sections below.

not a big deal: the diagram visualizes 6 steps. Also tagging @ricott1. Feel free to ignore this comment.

janhack replied:
I removed the exact number of steps.

ricott1 reviewed (approved):
Regarding the block storage stage, I think we should align on what is currently doing in the SDK, therefore I delegate the decision to @shuse2

janhack commented:
I added the block storage step to the description in LIP 60 and also updated the diagram.

AndreasKendziorra commented on proposals/lip-0058.md:
```diff
@@ -756,9 +730,9 @@ This function returns the slot number corresponding to the input timestamp.
+    if timestamp < genesisTimestamp:
         raise Exception('Input timestamp cannot be smaller than genesis block timestamp.')
-    elapsedTime = timestamp - genesisDataStore.timestamp
+    elapsedTime = timestamp - genesisTimestamp
```
oh, that's something I realized just now after having a look at LIP 0063 again. The timestamp of the genesis block could be any timepoint in its time slot. See [here](https://github.com/LiskHQ/lips/blob/ce9b653f7210f0d997a599486243ec0d031ed30b/proposals/lip-0063.md?plain=1#L332). Therefore, this calculation is not necessarily doing what it intended here. We could fix this here or in LIP 0063. I'm open for both.

For `getGeneratorAtTimestamp` I'm not sure if this is an issue. I guess not.
